### ISSUE TITLE
Inconsistent speeds during SFTP tranfers

### ISSUE THREAD

**Issue by nfebe** [PermanentOrg/sftp-service#220]
Originally reported by @danfuzz

> 
> I tried to upload another bunch of files yesterday via SFTP and ran into some trouble which I think is worth
> reporting:
> 
> During the initial part of the upload, my upload speed was super-slow (something like 20KB/sec)
> including stalling out repeatedly. After half an hour or so it seemed to stop having trouble (and
> accepted my traffic at more like 5MB/sec). Based on other activity on my local network, it seems likely
> that the issue was likely on the Permanent side.

**Comment by slifty** [comment]
Our current theory is that this is related to a compounding of two known issues:

1. Permanent does not expose the originally uploaded file associated with an archive record until AFTER the archive record is fully processed.  SFTP requires access to the original file to do things like return file statistics and file data.  

Ultimately this means that SFTP will become much slower than bandwidth allows for any file that takes a long time to process.  The plan that the Permanent team is currently exploring is to update the back end processing code so that the original file is populated much earlier in the pipeline.

2. Permanent has a known issue where processing of `.doc` and `.pdf` files can deadlock due to limitations with `libreoffice`.

Both of these independently are issues in themselves, but I think the first one is the most direct issue for the sftp service, and resolving that should improve performance for ALL file types (and obviate the second issue for SFTP use cases).

**Comment by danfuzz** [comment]
>processing of .doc and .pdf files can deadlock / slow down significantly

FWIW, at least in my case this wasn't part of the observed issue. (I'm pretty sure all the files I was uploading at the time were plain text files and `.tar.gz` files.)

**Comment by slifty** [comment]
@danfuzz very good to know -- if you're OK sharing, do you have a sense of whether the .tar.gz files were particularly large?

Either way I'd still suspect that the above processing sequence is at least responsible for SOME of the slowness, even if there may also be other issues at play here!).  There's also just the issue that uploads are waiting on network latency on a series of API calls to resolve.

**Comment by danfuzz** [comment]
@slifty IIRC the tarfiles in question ranged from something like 20 to 300 megs apiece.

**Comment by slifty** [comment]
@danfuzz It's been a while, but wanted to let you know that we made some changes last month which should dramatically improve SFTP speeds (we haven't done robust testing, but I've seen 2 MB / s regularly).

(1) We switched to a new, faster endpoint for loading folders.  This endpoint also provides all of the data SFTP needs which means we no longer need to make extra API calls
(2) We fixed a logical error that would bypass cached checks / would re-load folders unnecessarily when writing files.

I'm going to close this issue for now, and of course we can open a new issue if there are additional speed concerns.

**Comment by danfuzz** [comment]
Thanks for the update!
